crate/cli/src/actions/google/gmail_client/client_builder.rs

```diff
@@ -0,0 +1,117 @@
+use std::path::PathBuf;
+
+use reqwest::{Client, Response};
+use serde::Deserialize;
+
+use super::{service_account::ServiceAccount, token::retrieve_token, GoogleApiError};
+use crate::error::CliError;
+
+#[derive(Deserialize)]
+pub struct RequestError {
+    pub error: ErrorContent,
+}
+
+#[derive(Deserialize)]
+pub struct ErrorContent {
+    pub message: String,
+}
+
+#[derive(Debug, Clone)]
+struct GmailClientBuilder {
+    service_account: ServiceAccount,
+    user_id: String,
+}
+
+impl GmailClientBuilder {
+    pub fn new(conf_path: &PathBuf, user_id: &str) -> Result<Self, CliError> {
+        let service_account = ServiceAccount::load_from_config(conf_path)?;
+        Ok(Self {
+            service_account,
+            user_id: user_id.to_owned(),
+        })
+    }
+
+    pub async fn build(self) -> Result<GmailClient, CliError> {
+        let token = retrieve_token(&self.service_account, &self.user_id).await?;
+
+        Ok(GmailClient {
+            client: Client::new(),
+            token,
+            base_url: "https://gmail.googleapis.com/gmail/v1/users/".to_string() + &self.user_id,
+        })
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct GmailClient {
+    client: Client,
+    token: String,
+    base_url: String,
+}
+
+impl GmailClient {
+    pub async fn new(conf_path: &PathBuf, user_id: &str) -> Result<GmailClient, CliError> {
+        GmailClientBuilder::new(conf_path, user_id)?.build().await
+    }
+
+    pub async fn handle_response(response: Response) -> Result<(), CliError> {
+        if response.status().is_success() {
+            println!(
+                "{}",
+                response
+                    .text()
+                    .await
+                    .map_err(|e| GoogleApiError::ReqwestError(e.to_string()))?
+            );
+            Ok(())
+        } else {
+            let json_body = response
+                .json::<RequestError>()
+                .await
+                .map_err(|e| GoogleApiError::ReqwestError(e.to_string()))?;
+            Err(CliError::GmailApiError(json_body.error.message.to_string()))
+        }
+    }
+
+    pub async fn get(&self, endpoint: &str) -> Result<Response, GoogleApiError> {
+        Ok(self
+            .client
+            .get(self.base_url.to_string() + endpoint)
+            .bearer_auth(&self.token)
+            .send()
+            .await?)
```

Comment:
You can avoid a string clone, and use a `map_err`:
```suggestion
        self.client
            .get([&self.base_url, endpoint].concat())
            .bearer_auth(&self.token)
            .send()
            .await
            .map_err(GoogleApiError::from)
```